opm/io/eclipse/OutputStream.hpp

```diff
@@ -0,0 +1,175 @@
+/*
+  Copyright (c) 2019 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_IO_OUTPUTSTREAM_HPP_INCLUDED
+#define OPM_IO_OUTPUTSTREAM_HPP_INCLUDED
+
+#include <ios>
+#include <memory>
+#include <string>
+
+namespace Opm { namespace ecl {
+
+    class EclOutput;
+
+}} // namespace Opm::ecl
+
+namespace Opm { namespace ecl { namespace OutputStream {
+
+    struct Formatted { bool set; };
```

Comment:
My general opinion is that *as a creator/developer* of a code I enjoy creatinng "all kinds of types" - but as a reader of code it annoys me. This api:
```
Restart(ResultSet rset, bool fmt, bool unif);
```
is much easier to read than:
```
Restart(ResultSest, const Formatted& fmt, const Unified& unif)
```
and the fact that your approach can protect against argument order bugs is not enough in my book.

Reply:
I'll concede that the former may be easier to read if you're implementing the function, but I'm going to claim that the call site/client code becomes clearer in the second case.  Consider for instance
```C++
Restart rst(rset, true, false);
```
versus
```C++
Restart rst(rset, Formatted{ true }, Unified{ false });
```
The first requires somehow looking up the declaration (and possibly definition) of the `Restart` constructor while the second explicitly communicates the meaning of the flags we're setting.  As a side note, Qt uses similar practices in their APIs&mdash;although mostly in the form of strong `enum`s.  Personally, I've come to view unadorned `bool`s in public APIs with a lot of scepticism, especially if there are multiple such parameters and they're next to each other (or very close).

Reply:
We can discuss this until hell freezes over; I agree that the callsite you have illustrated:
```
Restart rst(rset, Formatted{ true }, Unified{ false });
```
looks nice - however when I come come to the declaration
```
Restart(RestartSet set, const Formatted& fmt, const Unified& unif);
```
and have to go searching for the declaration of `Formatted`and `Unified` I get annoyed. Using `const &` for those is further misleading - then I think this is some (moderately ....) large object which is owned elsewhere - when it turns it is just a pimped `bool`.

> I've come to view unadorned bools in public APIs with a lot of scepticism,

Technically this is of course a public api - but that is only in the name - it is deeply integrated in the `flow` codebase. For the codebase in question the callsite will typically be:
```
... ioconfig.getFormatted( ), ioconfig.getUnified( ));
```
i.e. no naked bools will be employed at the callsite.

I have made my point - do as you like.

Reply:
> For the codebase in question the callsite will typically be:
>
> ```
> ... ioconfig.getFormatted( ), ioconfig.getUnified( ));
> ```
> i.e. no naked bools will be employed at the callsite.

Even then the added wrapping is beneficial because it might cause the reader (and writer) to pause at cases like
```C++
Restart rstFile(rset, Formatted{ ioconfig.getUnified() }, Unified{ ioconfig.getUnified() });
```

> I have made my point - do as you like.

I'm going to continue using the wrappers, at least for the time being. If they cause too much trouble I'm open to unwrapping the `bool`s at a later point.